src/main/scala/com/microsoft/azure/cosmosdb/spark/ClientConfiguration.scala

```diff
@@ -0,0 +1,179 @@
+  def apply(config: Config): ClientConfiguration = {
+    val database : String = config.get(CosmosDBConfig.Database).get
+    val collection : String = config.get(CosmosDBConfig.Collection).get      
+    val authConfig : AuthConfig = validateAndCreateAuthConfig(config, database, collection)
+    val connectionPolicySettings : ConnectionPolicySettings = createConnectionPolicySettings(config)
+    val bulkExecutorSettings : BulkExecutorSettings = createBulkExecutorSettings(config)
+
+    // Get consistency level
+    val consistencyLevel = config.getOrElse(
+            CosmosDBConfig.ConsistencyLevel,
+            CosmosDBConfig.DefaultConsistencyLevel)
+
+    ClientConfiguration(
+      config.get(CosmosDBConfig.Endpoint).get,
+      authConfig,
+      connectionPolicySettings,
+      consistencyLevel,
+      database,
+      collection,
+      bulkExecutorSettings)
+  }
+
+  private def validateAndCreateAuthConfig(config: Config, database: String, collection: String) : AuthConfig = {
+    val resourceToken = config.getOrElse[String](CosmosDBConfig.ResourceToken, "")
+    val masterKey = config.getOrElse[String](CosmosDBConfig.Masterkey, "")
+
+    if ((resourceToken.isEmpty && masterKey.isEmpty) ||
+        (!resourceToken.isEmpty && !masterKey.isEmpty))
+    {
+        throw new IllegalArgumentException(
+            s"Configuration options '${CosmosDBConfig.Masterkey}' and " +
+            s"'${CosmosDBConfig.ResourceToken}' are mutually exclusive. " +
+            s"Exactly one of them must be defined.")    
+    }
+
+    // Check if resource token exists
+    if(resourceToken.isEmpty) {
+      AuthConfig(masterKey, None)
+    } else {
+      AuthConfig(resourceToken, Some(getCollectionLink(database, collection))) 
+    }
+  }
+
+  private def createBulkExecutorSettings(config: Config) : BulkExecutorSettings = {
+    val pkDef: Option[String] = config.get[String](CosmosDBConfig.PartitionKeyDefinition)
+    val maxMiniBatchImportSizeKB: Int = config
+      .getOrElse(CosmosDBConfig.MaxMiniBatchImportSizeKB, CosmosDBConfig.DefaultMaxMiniBatchImportSizeKB)
+    val maxMiniBatchUpdateCount: Int = config
+      .getOrElse(CosmosDBConfig.MaxMiniBatchUpdateCount, CosmosDBConfig.DefaultMaxMiniBatchUpdateCount)
+    val maxThroughputForBulkOperations = config
+      .get(CosmosDBConfig.WriteThroughputBudget)
+
+    BulkExecutorSettings(
+      maxMiniBatchUpdateCount,
+      maxMiniBatchImportSizeKB,
+      maxThroughputForBulkOperations,
+      pkDef)
+  }
+
+  private def createConnectionPolicySettings(config: Config) : ConnectionPolicySettings = {
+    val connectionMode = config.getOrElse(
+        CosmosDBConfig.ConnectionMode,
+        CosmosDBConfig.DefaultConnectionMode)
+    
+    val applicationName: String = config.getOrElse[String](CosmosDBConfig.ApplicationName, "")
+    val userAgentString: String = if (applicationName.isEmpty) {
+      s"${Constants.userAgentSuffix} ${ManagementFactory.getRuntimeMXBean.getName}"
```

Comment:
is it possible to suffix userAgentString with synapse when the connector is running in syanpse?

Reply:
As discussed it would be possible by passing down a config entry "application_name" (see CosmosDBConfig.scala- ApplicationName) with a Synapse prefix (if the user provided a value a combination of the user's application name and the Synapse prefix).